Add debugPrintKeyboardEvents flag

bleroux opened the pull request:
## Description

This PR adds a new debug flag named `debugPrintKeyboardEvents` to help debugging keyboard issues.

Keyboard code has some useful asserts but sometimes an assertion failure is related to the handling of previous key events. This debug flag will help understanding the flow of key events which leads to an assertion failure.

## Related Issue

Fixes https://github.com/flutter/flutter/issues/125627

## Tests

Adds 1 test.

dkwingsmt commented on packages/flutter/lib/src/services/hardware_keyboard.dart:
```diff
@@ -16,6 +16,49 @@ export 'package:flutter/foundation.dart' show DiagnosticPropertiesBuilder;
+]) {
+  if (kReleaseMode) {
+    throw StateError(
+      '_keyboardDebug was called in Release mode. It should always be wrapped in '
+      'an assert. Always call _keyboardDebug like so:\n'
```
```suggestion
      '_keyboardDebug was called in Release mode, which means they are called without being wrapped in '
      'an assert. Always call _keyboardDebug like so:\n'
```

dkwingsmt commented on packages/flutter/lib/src/services/hardware_keyboard.dart:
```diff
@@ -16,6 +16,49 @@ export 'package:flutter/foundation.dart' show DiagnosticPropertiesBuilder;
+/// Can be used to debug keyboard issues: each time a key event is received on
+/// the framework side, the event details and the current pressed state will
+/// be printed.
+bool debugKeyboardEvents = false;
```
Maybe `debugPrintKeyboardEvents`? See `packages/flutter/lib/src/gestures/debug.dart`

dkwingsmt reviewed (commented):
Probably want to add something similar to `debugAssertAllGesturesVarsUnset`.

Also I think `_keyboardDebug` might not be necessary, since similar functions are not used for `debugPrintHitTestResults`. It's just an `if` inside an `assert`.

bleroux commented:
> Probably want to add something similar to `debugAssertAllGesturesVarsUnset`.

Great, will do tomorrow.

> Also I think `_keyboardDebug` might not be necessary, since similar functions are not used for `debugPrintHitTestResults`. It's just an `if` inside an `assert`.

I took my inspiration from `debugFocusChanges` (`packages/flutter/lib/src/widgets/focus_managerdart`) implementation. It might be overkill for the moment (it could make sense if we use it in many places).

dkwingsmt commented:
> I took my inspiration from debugFocusChanges (packages/flutter/lib/src/widgets/focus_managerdart) implementation. It might be overkill for the moment (it could make sense if we use it in many places).

That's fair. I don't have strong opinion on this. :)

bleroux committed:
Add debugPrintKeyboardEvents flag

bleroux committed:
Update error message per review comment

bleroux commented:
I renamed the flag to `debugPrintKeyboardEvents` and moved its declaration to `packages/flutter/lib/src/services/debug.dart`.

> Probably want to add something similar to `debugAssertAllGesturesVarsUnset`.

As per the move to `/services/debug.dart`, I added a check in `debugAssertAllServicesVarsUnset`

dkwingsmt reviewed (approved):
LGTM
